Get table data based on labels #352

etantg opened the pull request:
[Get table data based on labels #352](https://github.com/bxservice/idempiere-rest/issues/352)
[Append label data to retrieved records #353](https://github.com/bxservice/idempiere-rest/issues/353)

Changes:
- Added a new query option: **$label** to serve as a label filter
- Added support for nested **$label** within the **$expand**
- Added a new query option: **showlabel**

Notes:
- **$label** query option is used to filter labels (i.e. AD_Label) within the assigned labels
`/api/v1/models/C_BPartner?$label=Name eq '%23Customer'` to get business partner records where the name of the Assigned Label > Label is '#Customer'.
- By default, **showlabel** query option returns a comma-separated list of assigned label names
`"assigned-labels": "#Customer, #Vendor"`
- **showlabel=list** query option returns a list of assigned label objects
```
"assigned-labels": [
	{
		"id": 1000000,
		"uid": "6d550225-2e0a-4203-9b82-0f2634f7d079",
		"AD_Client_ID": {
			"propertyLabel": "Tenant",
			"id": 11,
			"identifier": "GardenWorld",
			"model-name": "ad_client"
		},
		"AD_Org_ID": {
			"propertyLabel": "Organization",
			"id": 0,
			"identifier": "*",
			"model-name": "ad_org"
		},
		"Created": "2025-04-11T13:11:54Z",
		"CreatedBy": {
			"propertyLabel": "Created By",
			"id": 100,
			"identifier": "SuperUser",
			"model-name": "ad_user"
		},
		"Updated": "2025-04-11T13:11:54Z",
		"UpdatedBy": {
			"propertyLabel": "Updated By",
			"id": 100,
			"identifier": "SuperUser",
			"model-name": "ad_user"
		},
		"IsActive": true,
		"Name": "#Customer",
		"model-name": "ad_label"
	},
	{
		"id": 1000001,
		"uid": "87ae21d5-f06a-4cd8-977c-86712cf58da1",
		"AD_Client_ID": {
			"propertyLabel": "Tenant",
			"id": 11,
			"identifier": "GardenWorld",
			"model-name": "ad_client"
		},
		"AD_Org_ID": {
			"propertyLabel": "Organization",
			"id": 0,
			"identifier": "*",
			"model-name": "ad_org"
		},
		"Created": "2025-04-11T13:13:40Z",
		"CreatedBy": {
			"propertyLabel": "Created By",
			"id": 100,
			"identifier": "SuperUser",
			"model-name": "ad_user"
		},
		"Updated": "2025-04-11T13:27:37Z",
		"UpdatedBy": {
			"propertyLabel": "Updated By",
			"id": 100,
			"identifier": "SuperUser",
			"model-name": "ad_user"
		},
		"IsActive": true,
		"Name": "#Vendor",
		"model-name": "ad_label"
	}
]
```

hengsin commented:
hi @etantg , please add the new feature to https://wiki.idempiere.org/en/REST_Web_Services

etantg commented:
Updated the https://wiki.idempiere.org/en/REST_Web_Services

hengsin commented on com.trekglobal.idempiere.rest.api/src/com/trekglobal/idempiere/rest/api/v1/resource/impl/ModelResourceImpl.java:
```diff
@@ -116,9 +116,14 @@ protected ModelResourceImpl restView() {
 	}
 	
-	@Override
+	@Deprecated
```
don't have to deprecate this, please restore back the original override tag.

etantg replied:
Removed the deprecated annotation

hengsin commented on com.trekglobal.idempiere.rest.api/src/com/trekglobal/idempiere/rest/api/v1/resource/impl/ModelResourceImpl.java:
```diff
@@ -297,9 +308,15 @@ public Response getModels(String filter) {
 	}
 
-	@Override
+	@Deprecated
```
don't have to deprecate this, please restore back the original override tag.

hengsin commented on com.trekglobal.idempiere.rest.api/src/com/trekglobal/idempiere/rest/api/v1/resource/impl/ViewResourceImpl.java:
```diff
@@ -59,20 +59,32 @@ private ModelResourceImpl restView() {
 	}
 	
-	@Override
+	@Deprecated
```
don't have to deprecate this, please restore back the original override tag.

hengsin commented on com.trekglobal.idempiere.rest.api/src/com/trekglobal/idempiere/rest/api/v1/resource/impl/ViewResourceImpl.java:
```diff
@@ -59,20 +59,32 @@ private ModelResourceImpl restView() {
 	}
 
-	@Override
+	@Deprecated
```
don't have to deprecate this, please restore back the original override tag.

hengsin reviewed (approved):
looks good

hengsin commented:
> Hi @etantg / @hengsin / @PeterTakacs300 / @norbertbede
> 
> I tested this successfully, but I have a concern, not sure where it must be solved.
> 
> In the current way the user can define labels containing commas (also quotes)
> 
> So, in my tests I created the following label "this, is, crazy" and assigned to a BP
> 
> The result of the "showlabel" flag is this:
> 
> ```
> "assigned-labels": "#Customer, \"even quotes\", The best class, this, is, crazy, VIP"
> ```
> 
> I think it would be better to return an array of labels instead? something like:
> 
> ```
> "assigned-labels": [
> "#Customer",
> "\"even quotes\"",
> "The best class",
> "this, is, crazy",
> "VIP"
> ]
> ```
> 
> What do you think?

Yes, array is better.

etantg commented:
Modified the `showlabel` query option to follow the `$select` approach. 

- `showlabel` to include all columns

- `showlabel=name` to include just name

```
"assigned-labels": [
    "#Customer",
    "#Vendor"
]
```

- `showlabel=name,description` to include name and description

```
"assigned-labels": [
    {
        "Name": "#Customer",
        "Description": "Customers"
    },
    {
        "Name": "#Vendor",
        "Description": "Vendors"
    }
]
```
Updated the https://wiki.idempiere.org/en/REST_Web_Services

CarlosRuiz-globalqss reviewed (approved):
tested successfully
